SVD Lecture II

sladenheim opened the pull request:
This PR adjusts the content of the SVD II lecture. It now focuses on dimensionality reduction and introduces tSNE.

Closes #93

trgardos commented:
Care to comment on changes made to SVD I and kNN lectures?

sladenheim commented:
I noticed on the course notes webpage that kNN was producing a SyntaxWarning (https://tools4ds.github.io/DS701-Course-Notes/15-Classification-II-kNN.html). The minor change I made fixed it. It's because the latex inside of plots needs double \\.

I changed the SVD1 lecture based on your comments in issue #143 to make the rank r match rank k. 

I can always make more atomic PRs so that we don't group things together. I just figured I could fix those with this PR.

trgardos commented on ds701_book/jupyter_notebooks/10-Low-Rank-and-SVD.ipynb:
```diff
@@ -29,6 +29,96 @@
         "::::\n",
```
<p>Hmm... I wonder why this is showing as a change from your PR when it is already merged?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tools4ds/DS701-Course-Notes/pull/145/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

sladenheim replied:
<p>Not sure. I think that it might be because this branch was merged into main before these changes. No I am trying to merge the latest changes back in. Are you opposed to merging this in? I can always try to interactively rebase this, but it is going to take me some time to do that.</p>

trgardos replied:
Oh. It seems NBReview is confused in this case. The PR diff page shows just the rank index changes as you say ,but NBReview shows a bunch of changes that I made previously and already merged. We can disregard this comment.

trgardos replied:
Oh. I see. Not that's fine.

trgardos commented on ds701_book/jupyter_notebooks/15-Classification-II-kNN.ipynb:
```diff
@@ -36,19 +36,23 @@
       "cell_type": "markdown",
```
<p>hmm... strange. this seems to be from my earlier PR that is already merged. Why is it showing as a difference here? Perhaps if you didn't make changes here then you can remove these 2 other notebooks from this PR.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tools4ds/DS701-Course-Notes/pull/145/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

trgardos replied:
Same thing here. The PR diff shows just a minor change to the plot text whereas NBReview is showing still all the changes I originally made. We can disregard this too.

trgardos commented on ds701_book/jupyter_notebooks/11-Dimensionality-Reduction-SVD-II.ipynb:
```diff
@@ -5,259 +5,259 @@
       "metadata": {},
```
<p>Perhaps add some summary remarks about what we see from the plot?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tools4ds/DS701-Course-Notes/pull/145/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

trgardos commented on ds701_book/jupyter_notebooks/11-Dimensionality-Reduction-SVD-II.ipynb:
```diff
@@ -5,259 +5,259 @@
       "metadata": {},
```
<p>"SVD of the data matrix" or SVD of product of the transpose of the data matrix and data matrix?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tools4ds/DS701-Course-Notes/pull/145/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

sladenheim replied:
<p>You would take the SVD of X, the eigenvalue computation with X^TX is not always numerically stable (you're squaring the condition number of the data matrix X). The vectors V are the principal components and then you can eigenvalues from \Sigma^2.</p><p> </p><p>I will add a remark to clarify this.</p>

trgardos commented on ds701_book/11-Dimensionality-Reduction-SVD-II.qmd:
```diff
@@ -1,857 +1,759 @@
+Input: $\mathbf{x}_1,\ldots, \mathbf{x}_m$ with  $\mathbf{x}_i \in \mathbb{R}^n \: \: \forall \: i \in \{1, \ldots, n\}.$
 
-In order to better understand how the SVD provides us with this optimal transformation, let's consider the following criterion for a $k$-dimensional transformation:
+Output: $\mathbf{y}_1,\dots, \mathbf{y}_m$  with  $\mathbf{y}_i \in \mathbb{R}^d \: \: \forall \: i \in \{1, \dots, n\}$. 
```
I'm confused by the indexing and dimensionality. Are the $x_i$ the column vectors of the matrix? Then aren't they in $R^m$?. And same for $y_i$? Maybe the indexing on the matrix equation below is wrong?

sladenheim replied:
Added clarifying remarks below.

trgardos commented on ds701_book/11-Dimensionality-Reduction-SVD-II.qmd:
```diff
@@ -1,857 +1,759 @@
 
-But it also turns out to have a strong statistical guarantee.
-:::
+![Image Credit @novembre2008genes](figs/PCA_genes_Europe.png)
```
And perhaps add "And for comparison a color coded map of western europe is added and the same color coding was applied to the same data samples by country of origin."

sladenheim replied:
Added.

trgardos commented on ds701_book/11-Dimensionality-Reduction-SVD-II.qmd:
```diff
@@ -1,857 +1,759 @@
-__Find the $k$-dimensional hyperplane that is "closest" to the points.__
-:::
+:::: {.incremental}
+- the first principal components of the data almost reproduce the map of Europe
```
So the first 2 PCs seem to correspond to latitude and longitude!!

sladenheim replied:
Added comment to the observations.

trgardos commented on ds701_book/11-Dimensionality-Reduction-SVD-II.qmd:
```diff
@@ -1,857 +1,759 @@
 
 :::: {.fragment}
-In fact, this criterion is a transformation that captures the maximum __variance__ in the data.
+Would a similar study in the USA be effective?
```
Is this just meant for class discussion? Or is their evidence supporting an answer. I assume it wouldn't be as effective.

sladenheim replied:
Yes, this was for class discussion. It would not be as effective because of immigration.

trgardos commented on ds701_book/11-Dimensionality-Reduction-SVD-II.qmd:
```diff
@@ -1,857 +1,759 @@
 
-:::: {.fragment}
-That is, the resulting $k$-dimensional dataset is the one with maximum variance.
+## PCA Steps
```
So the intent here is to now formally define the PCA steps? Maybe a little more setup saying that is what we are going to do. However, then we proceed with an example, right?

sladenheim replied:
Yes. I think the underlying tension is that I planned to explain a lot of this in words so the text is more slide like as opposed to more of a book.

sladenheim replied:
I added some clarifying remarks that this was an outline of what is done, but that a more formal procedure to do this will be provided.

trgardos reviewed (changes requested):
Apologies for the many comments. I think it needs a bit more clarifications to make it more accessible.

sladenheim commented on ds701_book/11-Dimensionality-Reduction-SVD-II.qmd:
```diff
@@ -1,857 +1,759 @@
-:::: {.fragment}
-In other words, what rank $k$ matrix $X^{(k)} \in \mathbb{R}^{m\times k}$ is closest to $\tilde{X}$?
-::::
+We can plot the explained variance ratio and the total explained variance ratio.
```
This is related to T and \lambda_1/T on slide 17.

trgardos replied:
And where is "explained variance" defined?

sladenheim commented on ds701_book/11-Dimensionality-Reduction-SVD-II.qmd:
```diff
@@ -1,857 +1,759 @@
-[Source](https://liorpachter.wordpress.com/2014/05/26/what-is-principal-component-analysis/)
-:::
-::::
+Every real symmetric matrix $S$ has the factorization $V\Lambda V^{T}$, where $\Lambda$ is a diagonal matrix that contains the eigenvalues of S and teh columns of $V$ are orthogonal eigenvectors of $S$.
```
I wanted to use V here because it is more useful in highlighting the link to SVD. I picked X in the linear algebra refresher because it is commonly used in linear algebra textbooks. However, since we change to data matrices X, it might be something to re-think.

Are you ok with keeping it as is for now?

sladenheim replied:
Fixed the typo.

trgardos replied:
Yes. That is fine.

sladenheim commented on ds701_book/11-Dimensionality-Reduction-SVD-II.qmd:
```diff
@@ -1,857 +1,759 @@
+---
 
-Now, regardless of how we shift the green line, a right triangle is formed because the projection is orthogonal. 
+The covariance matrix $S \in \mathbb{R}^{n\times n}$ can be written as $S = V\Lambda V^T$ where
```
Yes. I say that in the previous slide that  every real symmetric matrix $S$ has the factorization $V\Lambda V^{T}$

sladenheim commented on ds701_book/11-Dimensionality-Reduction-SVD-II.qmd:
```diff
@@ -1,857 +1,759 @@
-:::
+---
+
+We can also plot the data in our transformed space.
```
I will change it. I don't think it is a good illustration of what we wanted to do.

sladenheim commented on ds701_book/11-Dimensionality-Reduction-SVD-II.qmd:
```diff
@@ -1,857 +1,759 @@
+SNPs describe the changes in DNA from a common base pair (A,T,C,G). A value of 0 means no changes to the base pair, a value of 1 means 1 change to the base pair, and a value of 2 means both base pairs change.
 
-![](figs/pca_figure1.jpeg){fig-align="center" width="60%"}
+The data for each individual consisted of approximately 500k SNPs.
```
Added.

sladenheim commented on ds701_book/11-Dimensionality-Reduction-SVD-II.qmd:
```diff
@@ -1,857 +1,759 @@
-:::: {.fragment}
-The vector $\bar{\mathbf{x}}^T$ is the centroid vector (mean vector) where the $i$-th entry is the average over the $i$-th row.
-::::
+2. Find the direction of the largest variance. This is the first principal component (PC1).
```
This is now explained in a later slide.

sladenheim commented on ds701_book/11-Dimensionality-Reduction-SVD-II.qmd:
```diff
@@ -1,857 +1,759 @@
-So for this case, let's construct the best 1-D approximation of the mean-centered data.
+# Scree plot (explained variance ratio)
+axs[0].plot(np.arange(1, len(pca.explained_variance_ratio_) + 1), pca.explained_variance_ratio_, marker='o', linestyle='--')
+axs[0].set_title('Scree Plot')
```
Defined.

sladenheim commented:
I addressed most of the comments except the ones that are unresolved. 

I removed the height and weight example and show a synthetic dataset that has one direction with a lot of variance and another with less. This illustrates the point I was trying to make much better. I also removed the figure in the transformed space since it is not that informative. They see the transformed data with the digits example.

Let me know what you think.

sladenheim commented on ds701_book/11-Dimensionality-Reduction-SVD-II.qmd:
```diff
@@ -1,857 +1,759 @@
+
+rng = np.random.default_rng(seed=42)
+
+height = rng.normal(loc=173, scale=6.35, size = 100)
```
Removed this example.
